src/client/common/terminal/environmentActivationProviders/condaActivationProvider.ts

```diff
@@ -81,13 +81,24 @@ export class CondaActivationCommandProvider implements ITerminalActivationComman
             if (versionInfo.minor >= CondaRequiredMinor) {
                 // New version.
                 const interpreterPath = await this.condaService.getInterpreterPathForEnvironment(envInfo);
-                const condaPath = await this.condaService.getCondaFileFromInterpreter(interpreterPath, envInfo.name);
-                if (condaPath) {
-                    const activatePath = path
-                        .join(path.dirname(condaPath), 'activate')
-                        .fileToCommandArgumentForPythonExt();
-                    const firstActivate = this.platform.isWindows ? activatePath : `source ${activatePath}`;
-                    return [firstActivate, `conda activate ${condaEnv.toCommandArgumentForPythonExt()}`];
+                const activatePath = await this.condaService.getActivationScriptFromInterpreter(
+                    interpreterPath,
+                    envInfo.name,
+                );
+                // eslint-disable-next-line camelcase
+                if (activatePath) {
+                    if (this.platform.isWindows) {
+                        return [activatePath, `conda activate ${condaEnv.toCommandArgumentForPythonExt()}`];
+                    }
+
+                    const condaInfo = await this.condaService.getCondaInfo();
+
+                    // eslint-disable-next-line camelcase
+                    if (condaInfo?.conda_shlvl === undefined || condaInfo.conda_shlvl === -1) {
+                        // conda may not be sourced correctly, so we need to source it.
+                        return [`source ${activatePath} ${condaEnv.toCommandArgumentForPythonExt()}`];
+                    }
+                    return [`conda activate ${condaEnv.toCommandArgumentForPythonExt()}`];
```

Comment:
Just realized simply sending `conda activate ${condaEnv.toCommandArgumentForPythonExt()}` will not work as `conda` may not be on `PATH`. Also `await this.condaService.getCondaInfo()` returns conda info for the global conda install but not pyenv specific one. For simplicity let us go with your original suggestion 🙂

```suggestion
                    // conda may not be sourced correctly, so we need to source it.
                    return [`source ${activatePath} ${condaEnv.toCommandArgumentForPythonExt()}`];
```

Reply:
if `conda` sourced correctly, The env var `CONDA_SHLVL` will be set. So we can use `conda activate <Env>`, otherwise use `source /path/to/activate <Env>`.   
BTW, if conda have been sourced successfully, `conda` is  a shell hook, not a binary. 

<img width="865" alt="图片" src="https://user-images.githubusercontent.com/16131917/165654698-7f310124-2bda-4de7-9193-71e6fcaac87f.png">

Reply:
We can check env var `CONDA_PYTHON_EXE` if same with current conda info? Do have any other ideas?

Reply:
I think always using `source ${activatePath} ${condaEnv.toCommandArgumentForPythonExt()}` should take care of it, regardless of whether conda is sourced or not (`conda init` has been done or not), what do you think?

Reply:
We should avoid relying on any internal details unless they're properly documented. In this case I don't see a need to do that.

> I think using conda activate <Env> is intuitive, because our users always use it in their interactive shell terminal.

If that is the case please try sending the command in two steps:
```
source ${activatePath}
conda activate ${condaEnv.toCommandArgumentForPythonExt()}
```
That way it's explicit that we're still using `conda activate <Env>`.

Reply:
> Oh right, thanks for clarifying. But we still have the pyenv conda issue:
> 
> ```
> const condaInfo = await this.condaService.getCondaInfo();
> ```
> 
> Only returns the conda info for global conda install. `conda info` for other installs may or may not have `condaInfo?.conda_shlvl` set, as hooks for each version of conda maybe different. Unless I'm missing something?

It's work well and I have tested it.  Hooks for each version of conda are same.

Reply:
> We should avoid relying on any internal details unless they're properly documented. In this case I don't see a need to do that.
> 
> > I think using conda activate  is intuitive, because our users always use it in their interactive shell terminal.
> 
> If that is the case please try sending the command in two steps:
> 
> ```
> source ${activatePath}
> conda activate ${condaEnv.toCommandArgumentForPythonExt()}
> ```
> 
> That way it's explicit that we're still using `conda activate <Env>`.

It's slower and a waste of time, because that conda will activate `base` first, then our `<Env>`.
 
_Originally posted by @MetRonnie in https://github.com/microsoft/vscode-python/issues/15818#issuecomment-1090186786_

Reply:
I see your point, but conda is free to modify/improve shell hooks, isn't it? There is no harm in sourcing the activate script every time. So I would suggest to simply go with that, we also have been doing that for Windows.

Reply:
> It's slower and a waste of time, because that conda will activate base first, then our <Env>.

Okay, in that case feel free to go with your solution, assuming `condaInfo?.conda_shlvl` is shared for all conda installs so we'll be good. If you could add a comment that documents it that would be great.

Reply:
I want try to find how to avoid source activatepath.

Reply:
@karrtikr In fact, the `condaService.getCondaInfo` always the global one, So `activatePath` will also in global path, since we use `condaInfo` to make it as absolute path.😂

Reply:
> So activatePath will also in global path, since we use condaInfo to make it as absolute path

That is actually not true, condaInfo is only used to calculate `activatePath` if we're unable to locate the activate script around the interpreter itself. For example for me the interpreter was `/home/kartik/.pyenv/versions/miniconda3-4.7.12/bin/python` and activate script was `/home/kartik/.pyenv/versions/miniconda3-4.7.12/bin/activate`, we didn't use `condaInfo` at all.

Reply:
> assuming condaInfo?.conda_shlvl is shared for all conda installs so we'll be good

If you can verify this value is set once any conda install is sourced, I think it's okay to rely on it to figure out when we need sourcing, otherwise please send the two commands in order.